### ISSUE TITLE
Cargo check 

### ISSUE THREAD

**Issue by otiv-willem-vanhulle** [bazelbuild/rules_rust#2978]
I am looking for an alternative to `cargo check --message-format=json` but cannot find it anywhere.

I have tried `bazel run @rules_rust//tools/upstream_wrapper:cargo check` but this expects a Cargo.toml file which I don't have.

How can I feed the right data to Rust Analyzer to check on save?

Minimal breaking example:
https://github.com/otiv-willem-vanhulle/bazel-rust-check

**Comment by havasd** [comment]
does this help?
`bazel build ... --output_groups=build_metadata --@rules_rust//:error_format=json`

**Comment by wvhulle** [comment]
@havasd yes I figured that out. But I have one last important question. How do the cargo check and Bazel build compare in terms of speed?

**Comment by havasd** [comment]
The commands are doing the same thing essentially.

we didn't do exact measurements but generally with cargo check the issue is that the caching is really broken for check and clippy. With bazel this is much better because of caching proper caching exact dependency tracking. If you combine it with remote cache and execution facilities than it is a different thing.

**Comment by havasd** [comment]
One thing is that based on what I see this requires pipelining otherwise the output group is not available.
`--@rules_rust//rust/settings:pipelined_compilation=True`

**Comment by wvhulle** [comment]
@havasd okay ! I noticed pipelined compilation fails for certain rare cases when used in combination with Clippy. Is this option really supposed to be used with Clippy?

**Comment by havasd** [comment]
If you can use clippy do you need separate cargo check? Clippy contains the cargo check as well. Or do you want to use for some faster lints?

Pipelined compilation can cause random failures because it requires reproducible crates in your project.
More details can be found here: https://github.com/bazelbuild/rules_rust/issues/1584

**Comment by otiv-willem-vanhulle** [comment]
@havasd, I run mostly Clippy. I can run the Bazel version of `cargo check` but I need for lint information from the Clippy groups. 

So I have 

```
build:clippy  --output_groups=+clippy_checks
```


Is there a benefit to using the following?

```
build:clippy   --@rules_rust//rust/settings:pipelined_compilation=True
```

This fails for very obscure reasons related to some kind of compiled crate version mismatch.

If the previous would work, is there an additional benefit to using ` --output_groups=build_metadata` like as follows?

```
build:clippy  --output_groups=build_metadata
```

**Comment by otiv-willem-vanhulle** [comment]
@havasd so there is no way to make the Clippy offered by Bazel rules Rust faster with extra options? Does it help to disable lints using options? It's hard to test because of the caching by Bazel.

**Comment by dragos** [comment]
Unfortunately, `rust_binary` does not declare `build_metadata` in `OutputGroupInfo` so this is a no-op for binary targets. Is there an alternative way to get a check-like functionality for binary targets?
